React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+    );
+
+    userEvent.click(getByText(buttonText));
+
+    expect(testOnClick).toBeCalledTimes(1);
+  });
+});
+```
+
+### Molecules test example
+
+Component:
+
+```tsx
+const UserCard: FC<UserCardProps> = ({ label }) => (
+  <Card>
+    <h3>{label}</h3>
+    <Button>click</Button>
+  </Card>
+);
+```
+
+Test:
+
+```jsx
+jest.mock("components/button");
+(Button as jest.Mock).mockReturnValue(<button />);
+
+describe("UserCard", () => {
+  it("Is rendering", () => {
+    const username = "Test User";
+
+    const { getByText } = render(<UserCard label={username} />);
+
+    expect(getByText(username)).toBeDefined();
+  });
+});
+```
+
+### Organism test example
+
+Component:
+
+```tsx
+const UserCardSection: FC<UserCardSectionProps> = ({ title }) => (
+  <Section>
+    <h1>{title}</h1>
+    <UserCard label="Test User" />
+  </Section>
+);
+```
+
+Test:
+
+```tsx
+jest.mock("components/user-card");
+(UserCard as jest.Mock).mockReturnValue(<div />);
+
+describe("UserCardSection", () => {
+  it("Is rendering", () => {
+    const title = "Test Title";
+
+    const { getByText } = render(<UserCardSection title={title} />);
+
+    expect(getByText(title)).toBeDefined();
+  });
+});
+```
+
+### Templates test example
+
+Component:
+
+```tsx
+export const UserTemplate: FC<UserTemplateProps> = (props) => (
+  <Grid {...props}>
+    <GridItem>
+      <Navbar />
+    </GridItem>
+    <GridItem>
+      <UserCardSection title="Test title" />
+    </GridItem>
+  </Grid>
+);
+```
+
+Test:
+
+```tsx
+jest.mock("components/user-card-section");
+(UserCardSection as jest.Mock).mockReturnValue(<div />);
+
+describe("UserTemplate", () => {
+  it("Is rendering", () => {
+    const { getByTestId } = render(
+      <UserTemplate data-testid="user-template-testid" />
+    );
+
+    expect(getByTestId("user-template-testid")).toBeDefined();
+  });
+});
+```
+
+## Page component test
+
+Pages tests should be located in `src/__tests__/pages` folder because Next.js is not allowing tests in `/pages` folder. Test name should be the same as the page file with `test.tsx` extension.
+
+Page example:
+
+```tsx
+const UserPage: NextPage = () => {
+  const { data, error } = useSWR<Array<User>, IResponseError>(
+    USER_KEY,
+    fetchUsers
+  );
+
+  if (error) {
+    return <ErrorPage />;
+  }
+
+  if (!data) {
+    return <LoadingPage />;
+  }
+
+  return <UserTemplate userList={data} />;
+};
+
+export default UserPage;
+```
+
+It this page example we should test all three states of the page component. To do that we should mock `<ErrorPage>`, `<LoadingPage>` and `<UserTemplate>` and check if it's rendered based on fetcher response.
+
+Test example:
+
+```tsx
+jest.mock("components/error-page");
+jest.mock("components/loading-page");
+jest.mock("components/user-template");
+jest.mock("fetchers/users");
+
+(ErrorPage as jest.Mock).mockReturnValue(
+  <div data-testid="error-page-testid" />
+);
+(LoadingPage as jest.Mock).mockReturnValue(
```

Comment:
If multiple pages depend on LoadingPage, would you extract this Jest mock to a separate file?

Also, if LoadingPage has some props, how would you test that Page component sends the correct values to LoadingPage via props?

Reply:
No, I think that each test should have its own mocks because they can be specific to that test and not sure if reusability has any benefit here.

Regarding the props I'm not sure, I'll do a research and add it to this example.